magicbot/magicrobot.py

```diff
@@ -433,31 +433,57 @@ def _setup_vars(self, cname, component):
         
         component_type = type(component)
         
+        # Iterate over variables with type annotations
+        for n, inject_type in getattr(component, '__annotations__', {}).items():
+            attr = getattr(component_type, n, True)
+
+            # If the variable is private or has a value associated
+            # hasattr will return True only if the variable has been initialized
+            if n.startswith('_') or hasattr(component_type, n):
+                # If the value given to the variable is an instance of a type and isn't a property
+                # raise an error. No double declaring types, e.g foo: type = type
+                if isinstance(attr, type) and not isinstance(attr, property):
+                    raise ValueError("Double Declaration: %s.%s has two type declarations" %(component_type .__name__, n))
+                continue
+
+            self._inject(n, inject_type, cname, component_type)
+
+        # Iterate over static variables
         for n in dir(component):
+            # If the variable is private or a proprty, don't inject
             if n.startswith('_') or isinstance(getattr(component_type, n, True), property):
                 continue
             
             inject_type = getattr(component, n)
             
+            # If the value assigned isn't a type, don't inject
             if not isinstance(inject_type, type):
                 continue
 
-            injectable = self._injectables.get(n)
-            if injectable is None:
-                if cname is not None:
-                    injectable = self._injectables.get('%s_%s' % (cname, n))
-
-            if injectable is None:
-                raise ValueError("Component %s has variable %s (type %s), which is not present in %s" %
-                                 (cname, n, inject_type, self))
-
-            if not isinstance(injectable, inject_type):
-                raise ValueError("Component %s variable %s in %s are not the same types! (Got %s, expected %s)" %
-                                 (cname, n, self, type(injectable), inject_type))
-
-            setattr(component, n, injectable)
-            self.logger.debug("-> %s as %s.%s", injectable, cname, n)
+            self._inject(n, inject_type, cname, component)
+
+    def _inject(self, n, inject_type, cname, component):
+        # Retrieve injectable object
+        injectable = self._injectables.get(n)
+        if injectable is None:
+            if cname is not None:
+                # Try for mangled names
+                injectable = self._injectables.get('%s_%s' % (cname, n))
+
+        # Raise error if injectable syntax used but no injectable was found.
+        # Most likely a staic variable was created but not declared as static
```

Comment:
This comment is… confusing. You can't have a "static variable" that's "not declared as static" (what does that mean?).